packages/liveblocks-react/src/factory.tsx

```diff
@@ -5,30 +5,54 @@ import type {
   History,
   Json,
   JsonObject,
-  Lson,
+  LiveObject,
   LsonObject,
   Others,
   Room,
   User,
 } from "@liveblocks/client";
-import { LiveList, LiveMap, LiveObject } from "@liveblocks/client";
-import type { Resolve, RoomInitializers } from "@liveblocks/client/internal";
-import { deprecate, errorIf } from "@liveblocks/client/internal";
+import { shallow } from "@liveblocks/client";
+import type {
+  Resolve,
+  RoomInitializers,
+  ToImmutable,
+} from "@liveblocks/client/internal";
+import { asArrayWithLegacyMethods } from "@liveblocks/client/internal";
 import * as React from "react";
+import { useSyncExternalStoreWithSelector } from "use-sync-external-store/shim/with-selector";
 
-import { useClient as _useClient } from "./client";
-import { useRerender } from "./hooks";
+import { useInitial, useRerender } from "./hooks";
 
 /**
- * "Freezes" a given value, so that it will return the same value/instance on
- * each subsequent render. This can be used to freeze "initial" values for
- * custom hooks, much like how `useState(initialState)` or
- * `useRef(initialValue)` works.
+ * For any function type, returns a similar function type, but without the
+ * first argument.
  */
-function useInitial<T>(value: T): T {
-  return React.useRef(value).current;
+type OmitFirstArg<F> = F extends (first: any, ...rest: infer A) => infer R
+  ? (...args: A) => R
+  : never;
+
+const noop = () => {};
+const identity: <T>(x: T) => T = (x) => x;
+
+const EMPTY_OTHERS =
+  // NOTE: asArrayWithLegacyMethods() wrapping should no longer be necessary in 0.19
+  asArrayWithLegacyMethods([]);
+
+function getEmptyOthers() {
+  return EMPTY_OTHERS;
 }
 
+export type MutationContext<
```

Comment:
Would it make sense to add "getMyPresence" or the current user in this context?

In the drawing board example, we often have to pass the selection as a dependency that comes from the presence.

```ts
const selection = useSelf(user => user.presence.selection);
return useMutation({ root } => { /*...*/ }, [selection]);
```

This invalidate a callback reference even when not strictly necessary. Something like this 👇 could be a slight improvement.

```ts
return useMutation({ root, self } => { /*...*/ }, []);
```

Reply:
Agreed, I’ve missed this in a couple places too, and wished it was there on the context for convenience! If we add this one in though, should we also include all other contexts? I’d be cheap to add those in I think? For example, [here](https://github.com/liveblocks/liveblocks/pull/384/files#diff-a154e780519dd9c827a011031e35442f7be80f0ed2f78ef1a83ab84c30728826R357) I needed access to the immutable storage value. I can aLeo imagine situations where you want to mutate something based on the state of `others`.

Would it make sense to include all of the following?

- `root`
- `setMyPresence`
- `me`
- `others`
- `immutableRoot` (equiv to `root.toImmutable()` so not strictly needed)